Switch to a max counterparty's `dust_limit_satoshis` constant

ariard opened the pull request:
This is a more conservative revamp of #575, see discussion there for rational. Contrary to what we previously discussed we don't have a risk of channel closure triggered by third-party as this check is enforced at channel opening. If our counterparty announces a `dust_limit_satoshis` above 660 sats, we halt the opening.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -444,6 +444,9 @@ pub const COMMITMENT_TX_WEIGHT_PER_HTLC: u64 = 172;
 /// it's 2^24.
 pub const MAX_FUNDING_SATOSHIS: u64 = 1 << 24;
 
+/// Maximum counterparty `dust_limit_satoshis` allowed. 2 * standard dust threshold on p2wsh output
```
Can you add some docs on how you derived the 330 part of this value? In Bitcoin Core I don't think there is a "standard p2wsh output dust threshold".

TheBlueMatt reviewed (commented):
Concept ACK

TheBlueMatt commented:
Note that `ln::channel::tests::channel_reestablish_no_updates` and `ln::channel::tests::test_holder_vs_counterparty_dust_limit` are currently failing.

ariard commented:
Fixed test at 3f534f8 by adding a default HOLDER_DUST_LIMIT_SATOSHIS, replacing the previous `derive_holder_dust_limit_satoshis`.

As this getter is function of current feerate, it doesn't guarantee to be compatible anymore with our new upper bound MAX_DUST_LIMIT_SATOSHIS. This might have lead to reject of our own channel opening.

This also removes the config option `min_limit_satoshis` to always requires that counterparty dust limit satoshis is between HOLDER_DUST_LIMIT_SATOSHIS and MAX_DUST_LIMIT_SATOSHIS. 

I think this needs another reviewer beyond Matt.

Note also I'll manually test the new 330 satoshis limit against Core to test that my computations are good. Should be easier now we have a sample.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -511,7 +522,7 @@ impl<Signer: Sign> Channel<Signer> {
 		}
 		let background_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Background);
-		if Channel::<Signer>::get_holder_selected_channel_reserve_satoshis(channel_value_satoshis) < Channel::<Signer>::derive_holder_dust_limit_satoshis(background_feerate) {
+		if Channel::<Signer>::get_holder_selected_channel_reserve_satoshis(channel_value_satoshis) < HOLDER_DUST_LIMIT_SATOSHIS {
```
The error message here shouldn't print anything about fees anymore - the only issue, I think, is if the total channel value is < 330, plus the fee lookup one line up can be dropped.

ariard replied:
Upgraded with a new message, right doesn't need feerate lookup anymore.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -687,11 +698,11 @@ impl<Signer: Sign> Channel<Signer> {
-		if msg.dust_limit_satoshis < config.peer_channel_config_limits.min_dust_limit_satoshis {
-			return Err(ChannelError::Close(format!("dust_limit_satoshis ({}) is less than the user specified limit ({})", msg.dust_limit_satoshis, config.peer_channel_config_limits.min_dust_limit_satoshis)));
+		if msg.dust_limit_satoshis < HOLDER_DUST_LIMIT_SATOSHIS {
+			return Err(ChannelError::Close(format!("dust_limit_satoshis ({}) is less than the user specified limit ({})", msg.dust_limit_satoshis, HOLDER_DUST_LIMIT_SATOSHIS)));
```
This should not mention "user specified" anymore - its not user-specified.

valentinewallace commented:
Hmm, I guess I'm just questioning the concept of this PR because not everyone's gonna be using bitcoind? (If I'm misunderstanding this, could someone outline the rationale more explicitly rather than a pointer to #575 ?) Plus, aren't we supposed to be the "flexible" lightning implementation? :stuck_out_tongue: I think I'm just missing something but just trying to understand.

TheBlueMatt commented:
> not everyone's gonna be using bitcoin

As to the minimum dust limit, Bitcoin Core's relay rules are effectively consensus for us. People can run alternate nodes, but if your transaction doesn't meet Bitcoin Core's relay rules, probably it wont find its way to a miner, and even if it did the miner would have to be running something other than Bitcoin Core.

Ultimately, this PR is about addressing lightning "dust inflation" - if your counterparty sets the dust limit higher than is necessary, they can send a number of HTLCs, leave them pending, and then close the channel, burning lots of your funds to fee. We assume that lightning counterparties aren't miners largely for this reason, but ideally it wouldn't be so trivial to burn funds.

valentinewallace reviewed (commented):
I'm ACK 7c25fb0526154b850826c46a9ac1b578023e051b mod fixing CI and addressing the other comments :)

ariard commented:
Updated with comments fixed at 0057afa. Main changes since last time is renaming HOLDER_DUST_LIMIT_SATOSHIS to MIN_DUST_LIMIT_SATOSHIS as this effectively a min required or setup by default on both holder/counterparty commitment transactions.

The only bound we don't enforce is MAX_DUST_LIMIT_SATOSHIS on our own commitment transactions, up to the counterparty to do it.

TheBlueMatt commented:
The full_stack_target fuzz failure here looks separate from the one fixed in #902 and I assume is new in the PR here.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -719,13 +730,12 @@ impl<Signer: Sign> Channel<Signer> {
 		let holder_selected_channel_reserve_satoshis = Channel::<Signer>::get_holder_selected_channel_reserve_satoshis(msg.funding_satoshis);
-		if holder_selected_channel_reserve_satoshis < holder_dust_limit_satoshis {
-			return Err(ChannelError::Close(format!("Suitable channel reserve not found. remote_channel_reserve was ({}). dust_limit_satoshis is ({}).", holder_selected_channel_reserve_satoshis, holder_dust_limit_satoshis)));
+		if holder_selected_channel_reserve_satoshis < MIN_DUST_LIMIT_SATOSHIS {
```
Since we select the `holder_selected_channel_reserve_satoshis`, could we just ensure that we never select a reserve below `MIN_DUST_LIMIT_SATOSHIS`?

ariard replied:
Do you mean instead of returning an API error, rouding up the `channel_reserve_satoshis` with MIN_DUST_LIMIT_SATOSHIS.

I think I prefer the user to swallow the error and having manually to bounce up the channel value instead of us doing it automatically. We might silently encroach on its expected liquidity ready to use and falsify higher application logic like an accounting app...No a strong opinion here.

valentinewallace replied:
It just seem like it doesn't make sense for `get_holder_selected_channel_reserve_satoshis` to ever return a value less than `330`. But, fine to leave that for follow-up

ariard replied:
I don't think our API prevent a user to try a `new_outbound` with less than 330 sat ? And if does so `get_holder_selected_channel_reserve_satoshis` will return the exact value. 

That said there is a TODO to make more sense of `get_holder_selected_channel_reserve_satoshis`. We can address it at that time.

valentinewallace reviewed (commented):
Just one question but this looks good

TheBlueMatt commented on fuzz/src/full_stack.rs:
```diff
@@ -917,6 +914,5 @@ mod tests {
 		assert_eq!(log_entries.get(&("lightning::ln::peer_handler".to_string(), "Handling UpdateHTLCs event in peer_handler for node 030200000000000000000000000000000000000000000000000000000000000000 with 1 adds, 0 fulfills, 0 fails for channel 3a00000000000000000000000000000000000000000000000000000000000000".to_string())), Some(&3)); // 7
 		assert_eq!(log_entries.get(&("lightning::ln::peer_handler".to_string(), "Handling UpdateHTLCs event in peer_handler for node 030000000000000000000000000000000000000000000000000000000000000002 with 0 adds, 1 fulfills, 0 fails for channel 3d00000000000000000000000000000000000000000000000000000000000000".to_string())), Some(&1)); // 8
 		assert_eq!(log_entries.get(&("lightning::ln::peer_handler".to_string(), "Handling UpdateHTLCs event in peer_handler for node 030000000000000000000000000000000000000000000000000000000000000002 with 0 adds, 0 fulfills, 1 fails for channel 3d00000000000000000000000000000000000000000000000000000000000000".to_string())), Some(&2)); // 9
-		assert_eq!(log_entries.get(&("lightning::chain::channelmonitor".to_string(), "Input spending counterparty commitment tx (0000000000000000000000000000000000000000000000000000000000000089:0) in 0000000000000000000000000000000000000000000000000000000000000074 resolves outbound HTLC with payment hash ff00000000000000000000000000000000000000000000000000000000000000 with timeout".to_string())), Some( ...
```
Why do we not hit this anymore? seems like this implies we now have less coverage?

TheBlueMatt reviewed (approved):
I'll happily fix up the fuzz test after merge, I think we shouldn't hold this up on it. Looks good otherwise.
